app/controllers/textings_controller.rb

```diff
@@ -0,0 +1,108 @@
+class TextingsController < ApplicationController
+
+  def create
+    # Make a request to Textellent API. If request successful we update our DB
+    response = post_textellent(get_db_data)
+
+    if response['status'] == 'success'
+      recipient = TextingRecipient.find_by(phone_number: texting_params[:phone_number])
+
+      if recipient
+        texting = Texting.new(
+          texting_recipient_id: recipient.id,
+          service_id: texting_params[:service_id]
+          )
+
+        if texting.save
+          render status: :ok, json: { message:'Message sent' }
+        else
+          render status: :bad_request
+        end
+
+      else
+        recipient = TextingRecipient.new(
+          recipient_name: texting_params[:recipient_name],
+          phone_number: texting_params[:phone_number]
+          )
+
+        recipient.save
+
+        texting = Texting.new(
+          texting_recipient_id: recipient.id,
+          service_id: texting_params[:service_id]
+          )
+
+        if texting.save
+          render status: :ok, json: { message:'Message sent' }
+        else
+          render status: :bad_request
+        end
+
+      end
+
+    else
+      render status: :bad_request, json: { error: response['message'] }
+    end
+  end
+
+  private
+
+  # Generate a data object to send to Textellent API
+  def get_db_data
+    recipient_name = texting_params[:recipient_name]
+    phone_number = texting_params[:phone_number]
+    service = Service.find(texting_params[:service_id])
+    category_ids = service.category_ids
+    resource = Resource.find(service.resource_id)
+    address = resource.addresses
+    phone = resource.phones[0].number
+
+    categories = []
+    category_ids.each do |category_id|
+      categories << Category.find(category_id).name
+    end
+
+    data = {
+    "firstName" => recipient_name,
+    "lastName" => "",
+    "mobilePhone" => phone_number,
+    "phoneAlternate": "",
+    "phoneHome" => "",
+    "phoneWork" => "",
+    "tags" => categories,
+    "engagementType" => "Resource Info",
+    "engagementInfo" => {
+        "Org_Name" => service.name,
+        "Org_Address1" => address[0].address_1,
```

Comment:
This is less common, but some resources do not have any addresses. This is usually because the resource has no physical location and are just some online resource, but sometimes this is also because the data is incomplete in the database. You may want to be careful that you don't trigger an exception when trying to index into a zero-length array.